sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala

```diff
@@ -66,8 +69,12 @@ private[sql] object JsonInferSchema {
                 s"Parse Mode: ${FailFastMode.name}.", e)
           }
         }
-      }
-    }.fold(StructType(Nil))(
+      }.fold(StructType(Nil))(
+        compatibleRootType(columnNameOfCorruptRecord, parseMode))
+      Iterator(typeInPartition)
+    }.collect()
```

Comment:
Previously two foldings with `compatibleRootType` on executors and local in one `RDD.fold` call all use default SQL config values. Now on executors SQL config can work, so inconsistency happens because local folding uses default SQL config.

Reply:
good catch! but wondering how the test passed in my PR...

Reply:
the collect here may cause OOM while the previous code doesn't have this problem. Let's revert my PR first and think about how to fix it.

Reply:
> good catch! but wondering how the test passed in my PR...

It is somehow flaky. If all types are folded at executor sides, when they are going to fold at local, it just merging `StructType()` and `StructType(StructField("id"), StructField("ID"))`. So you can still get current schema back.

But if unfortunately, you have one partition with only `id` column, you need to merge `StructType(StructField("id"))` and `StructType(StructField("ID")` in local. Then the problem will happen.

Reply:
> the collect here may cause OOM while the previous code doesn't have this problem. Let's revert my PR first and think about how to fix it.

We just iterate rows and fold a data type. So I think it shouldn't be OOM since we don't load all data at once?

Reply:
`RDD.fold` won't allocate an array for keeping the result. It merges the data types right after a task is finished and return a data type.

Reply:
Oh, I see. `RDD.fold` does folding immediately when a data type is returned from one task. But we now allocate an array to hold all data types.

I think we can use `RDD.toLocalIterator` instead of `collect`?